File: crypto/RSA.cpp
#include <iostream>
#include <vector>
#include <cmath>

long long public_key, private_key, n;

long long power(long long base, long long exponent, long long mod) {
    long long result = 1;
    while (exponent > 0) {
        if (exponent % 2 == 1) {
            result = (result * base) % mod;
        }
        exponent = exponent >> 1;
        base = (base * base) % mod;
    }
    return result;
}

long long encrypt(long long message) {
    return power(message, public_key, n);
}

long long decrypt(long long encrypted_text) {
    return power(encrypted_text, private_key, n);
}

std::vector<long long> encoder(const std::string& message) {
    std::vector<long long> encoded;
    for (char letter : message) {
        encoded.push_back(encrypt(static_cast<long long>(letter)));
    }
    return encoded;
}

std::string decoder(const std::vector<long long>& encoded) {
    std::string s;
    for (long long num : encoded) {
        s += static_cast<char>(decrypt(num));
    }
    return s;
}

void setkeys() {
    long long p, q, phi;

    std::cout << "Enter prime number p: ";
    std::cin >> p;
    std::cout << "Enter prime number q: ";
    std::cin >> q;

    std::cout << "Calculating modulus n = p * q..." << std::endl;
    n = p * q;
    std::cout << "Modulus n is set to: " << n << std::endl;

    phi = (p - 1) * (q - 1);

    std::cout << "Enter public key exponent e: ";
    std::cin >> public_key;

    std::cout << "Enter private key exponent d: ";
    std::cin >> private_key;
}

int main() {
    setkeys();

    std::cin.ignore(); // Ignore newline left in the input buffer
    std::cout << "\nEnter message: ";
    std::string message;
    std::getline(std::cin, message);

    auto coded = encoder(message);

    std::cout << "\nInitial message:\n" << message << std::endl;

    std::cout << "\nThe encoded message (encrypted by public key)\n";
    for (long long p : coded) {
        std::cout << p << " ";
    }

    std::string decrypted_message = decoder(coded);
    std::cout << "\n\nThe decoded message (decrypted by private key)\n" << decrypted_message << std::endl;

    return 0;
}

File: crypto/RSA.py
import random
import math

# Global variables for public and private keys, and modulus n
public_key = None
private_key = None
n = None

def encrypt(message):
    global public_key, n
    e = public_key
    encrypted_text = 1
    while e > 0:
        encrypted_text *= message
        encrypted_text %= n
        e -= 1
    return encrypted_text

def decrypt(encrypted_text):
    global private_key, n
    d = private_key
    decrypted = 1
    while d > 0:
        decrypted *= encrypted_text
        decrypted %= n
        d -= 1
    return decrypted

def encoder(message):
    encoded = []
    for letter in message:
        encoded.append(encrypt(ord(letter)))
    return encoded

def decoder(encoded):
    s = ''
    for num in encoded:
        s += chr(decrypt(num))
    return s

def setkeys():
    global public_key, private_key, n
    # Remove automatic prime generation and take inputs from the user
    p = int(input("Enter prime number p: "))
    q = int(input("Enter prime number q: "))
    
    n = p * q
    phi = (p - 1) * (q - 1)
    
    e = int(input("Enter public key exponent e: "))
    public_key = e
    
    d = int(input("Enter private key exponent d: "))
    private_key = d

# Set keys based on user input
setkeys()

# Input message from user
message = input("\n\nEnter message: ")
coded = encoder(message)

print("\n\nInitial message:")
print(message)

print("\n\nThe encoded message (encrypted by public key)\n")
print(''.join(str(p) for p in coded))

print("\n\nThe decoded message (decrypted by private key)\n")
print(decoder(coded))
print('\n\n')
